streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -222,21 +226,21 @@ public <KR> KStream<KR, V> selectKey(final KeyValueMapper<? super K, ? super V,
                                          final Named named) {
         Objects.requireNonNull(mapper, "mapper can't be null");
         Objects.requireNonNull(named, "named can't be null");
-
+        final boolean repartitionRequired = !(graphNode instanceof PartitionPreservingNode);
         final ProcessorGraphNode<K, V> selectKeyProcessorNode = internalSelectKey(mapper, new NamedInternal(named));
-        selectKeyProcessorNode.keyChangingOperation(true);
+        selectKeyProcessorNode.keyChangingOperation(repartitionRequired);
```

Comment:
`selectKey` should always set to `true` -- a downstream `markAsPartitioned()` might reset the flag to `false` but we don't know what happens downstream